MapleServer2/Types/Inventory.cs

```diff
@@ -136,6 +139,151 @@ public Inventory(long id, Dictionary<InventoryTab, short> extraSize, List<Item>
+        }
+    }
+
+    public void IncrementSetBonus(GameSession session, SetBonus setBonus)
+    {
+        setBonus.EquipCount++;
+
+        bool addedEffect = false;
+        bool recompute = false;
+
+        foreach (SetBonusMetadata bonus in setBonus.Bonuses.Parts)
+        {
+            if (bonus.Count != setBonus.EquipCount)
+            {
+                continue;
+            }
+
+            for (int i = 0; i < bonus.AdditionalEffectIds.Length; ++i)
+            {
+                int id = bonus.AdditionalEffectIds[i];
+                int level = bonus.AdditionalEffectLevels[i];
+
+                if (id != 0)
+                {
+                    session.Player.FieldPlayer?.AdditionalEffects.AddEffect(new(id, level));
+                }
+            }
+
+            addedEffect |= bonus.AdditionalEffectIds.Length > 0;
+            recompute |= bonus.AdditionalEffectIds.Length == 0;
+        }
+
+        if (recompute && !addedEffect)
+        {
+            session.Player.FieldPlayer?.ComputeStats();
+        }
+    }
+
+    public void ItemEquipped(GameSession session, Item item)
+    {
+        foreach (SetBonus setBonus in SetBonuses)
+        {
+            if (!setBonus.HasItem(item))
+            {
+                continue;
+            }
+
+            IncrementSetBonus(session, setBonus);
+
+            return;
+        }
+
+        SetBonus? newBonus = SetBonus.From(item);
+
+        if (newBonus is null)
+        {
+            return;
+        }
+
+        SetBonuses.Add(newBonus);
+
+        IncrementSetBonus(session, newBonus);
+    }
+
+    public void ItemUnequipped(GameSession session, Item item)
+    {
+        SetBonus? removeBonus = null;
+
+        foreach (SetBonus setBonus in SetBonuses)
+        {
+            if (!setBonus.HasItem(item))
+            {
+                continue;
+            }
+
+            removeBonus = setBonus;
+
+            break;
+        }
+
+        if (removeBonus is null)
+        {
+            return;
+        }
+
+        removeBonus.EquipCount--;
+
+        if (removeBonus.EquipCount == 0)
+        {
+            SetBonuses.Remove(removeBonus);
+        }
+
+        bool removedEffect = false;
+        bool recompute = false;
+
+        foreach (SetBonusMetadata bonus in removeBonus.Bonuses.Parts)
+        {
+            if (bonus.Count != removeBonus.EquipCount - 1)
+            {
+                continue;
+            }
+
+            for (int i = 0; i < bonus.AdditionalEffectIds.Length; ++i)
+            {
+                session.Player.FieldPlayer?.AdditionalEffects.GetEffect(bonus.AdditionalEffectIds[i])?.Stop(session.Player.FieldPlayer);
+            }
+
+            removedEffect |= bonus.AdditionalEffectIds.Length > 0;
+            recompute |= bonus.AdditionalEffectIds.Length == 0;
+
+            break;
+        }
+
+        if (recompute && !removedEffect)
+        {
+            session.Player.FieldPlayer?.ComputeStats();
+        }
+    }
```

Comment:
ItemEquipped  has multiple references from other places too. It is used twice for two different item types in RecomputeSetBonus, which is called on field enter. ItemUnequipped is only called in one spot, but the control flow is complex enough that I don't think it would cleanly fit into the function it is called in as is.

Maybe they could be renamed and removed from the interface? I originally put it there thinking more spots would need it, but never re-evaluated if I should remove it from there or not. I don't think either are used, while RecomputeSetBonuses is.